include/librealsense2/dds/dds-device.cpp

```diff
@@ -0,0 +1,92 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2022 Intel Corporation. All Rights Reserved.
+
+#include "dds-device.h"
+#include "dds-participant.h"
+#include "topics/device-info/device-info-msg.h"
+#include <librealsense2/utilities/easylogging/easyloggingpp.h>
+#include <fastdds/rtps/common/Guid.h>
+
+#include <map>
+
+using namespace eprosima::fastdds::dds;
+
+namespace librealsense {
+namespace dds {
+
+
+namespace {
+// The map from guid to device is global
+typedef std::map< dds::dds_guid, std::weak_ptr< dds::dds_device > > guid_to_device_map;
+guid_to_device_map guid_to_device;
+std::mutex devices_mutex;
+}  // namespace
+
+
+class dds_device::impl
+{
+public:
+    topics::device_info _info;
+    std::shared_ptr< dds::dds_participant > const _participant;
+
+    impl( std::shared_ptr< dds::dds_participant > const & participant,
+          dds::dds_guid const & guid,
+          dds::topics::device_info const & info )
+        : _info( info )
+        , _participant( participant )
+    {
+        (void)guid;
+    }
+};
+
+
+std::shared_ptr< dds::dds_device >
+dds_device::find_or_create_dds_device( std::shared_ptr< dds::dds_participant > const & participant,
+                                       dds::dds_guid const & guid,
+                                       dds::topics::device_info const & info,
+                                       bool create_it )
+{
+    std::weak_ptr< dds::dds_device > wdev;
+    std::shared_ptr< dds::dds_device > dev;
+
+    std::lock_guard< std::mutex > lock( devices_mutex );
+    auto it = guid_to_device.find( guid );
+    if( it != guid_to_device.end() )
+    {
+        dev = it->second.lock();
+        if( ! dev )
+        {
+            // The device is no longer in use; clear it out
+            guid_to_device.erase( it );
+        }
+    }
+    else if( create_it )
+    {
+        guid_to_device.emplace( guid, dev );
```

Comment:
I am not sure this works.
The order needs to be changed IMO.

```
dev = ...
guid_to_device.emplace( guid, dev );
```

Do you really see the map's dev being updated? I think it will be empty..
Maybe it works when you tested it cause you are holding the shared_ptr you returned?

Reply:
You're right this went thru some changes and this got mixed up somehow... bug :)

Reply:
👍